src/audio/OpenALAudioEngine.cpp

```diff
@@ -0,0 +1,491 @@
+
+            alBufferData(m_buffer,
+                         fmt == Format::Mono ? AL_FORMAT_MONO16 : AL_FORMAT_STEREO16,
+                         buf, len, samplingFreq);
+            error = alGetError();
+            if (error != AL_NO_ERROR)
+            {
+                alDeleteBuffers(1, &m_buffer);
+                throw OpenALException(error);
+            }
+        }
+
+        ~OpenALSound() override
+        {
+            if (m_hasSource)
+            {
+                // Stop source and detach from buffer
+                alSourceStop(m_source);
+                alSourcei(m_source, AL_BUFFER, 0);
+                m_engine->m_freeSources.push_back(m_source);
+            }
+            
+            if(m_hasBuffer) alDeleteBuffers(1, &m_buffer);
+        }
+
+        ALuint seizeSource()
+        {
+            if (!m_hasSource)
+            {
+                throw std::runtime_error("Sound has no source");
+            }
+
+            m_hasSource = false;
+            return m_source;
+        }
+
+        bool hasSource() const
+        {
+            return m_hasSource;
+        }
+
+        void pitch(float p) override
+        {
+            getSource();
+            alSourcef(m_source, AL_PITCH, p);
+            ALenum error = alGetError();
+            if (error != AL_NO_ERROR) throw OpenALException(error);
+        }
+
+        void gain(float g) override
+        {
+            getSource();
+            alSourcef(m_source, AL_GAIN, g);
+            ALenum error = alGetError();
+            if (error != AL_NO_ERROR) throw OpenALException(error);
+        }
+
+        void maxDistance(float md) override
+        {
+            getSource();
+            alSourcef(m_source, AL_MAX_DISTANCE, md);
+            ALenum error = alGetError();
+            if (error != AL_NO_ERROR) throw OpenALException(error);
+        }
+
+        void position(const Math::float3& p) override
+        {
+            getSource();
+            alSourcefv(m_source, AL_POSITION, reinterpret_cast<const ALfloat*>(&p));
+        }
+
+        void velocity(const Math::float3& v) override
+        {
+            getSource();
+            alSourcefv(m_source, AL_VELOCITY, reinterpret_cast<const ALfloat*>(&v));
+        }
+
+        void direction(const Math::float3& d) override
+        {
+            getSource();
+            alSourcefv(m_source, AL_DIRECTION, reinterpret_cast<const ALfloat*>(&d));
+        }
+
+        void relative(bool r) override
+        {
+            getSource();
+            alSourcei(m_source, AL_SOURCE_RELATIVE, r ? AL_TRUE : AL_FALSE);
+        }
+
+        void looping(bool l) override
+        {
+            getSource();
+            alSourcei(m_source, AL_LOOPING, l ? AL_TRUE : AL_FALSE);
+        }
+
+        State state() override
+        {
+            if (!m_hasSource) return State::Stopped;
+            ALint s;
+            alGetSourcei(m_source, AL_SOURCE_STATE, &s);
```

Comment:
nit: as it returns number, you can assign these numbers in enum and cast value of `s` to `State`.

Reply:
I'm not sure I particularly like this idea, because it ties it specifically to the states that OpenAL returns, which is the reason I wrote a wrapper around it...